Set password after saving Buddypress settings

Floppy opened the pull request:
### Changes

This change adds support for Buddypress password changes, which don't fire any of the existing hooked actions. Instead, we hook onto `bp_core_general_settings_after_save`.

I'm still testing it here to make sure it works, but I'm opening this PR early for suggestions and feedback. Error handling, for instance, is certainly a problem, as Buddypress doesn't seem to have a hook that allows errors to feed back.

### Testing

Testing needs a WordPress instance with Buddypress installed. When saving email or password from buddypress account settings, the Auth0 equivalents should also be updated. I don't know if there is a way I should be installing Buddypress to test it; there doesn't seem to be anything that does that for Woocommerce that I can see.

* [x] I included manual testing steps above, if applicable
* [x] This change adds unit test coverage
* [x] This change has been tested on the latest version of the platform/language or why not

### Checklist

* [x] I read [Auth0's general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)?
* [x] I read [Auth0's code of conduct guidelines](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
* [ ] All existing and new tests complete without errors
* [ ] All code quality tools/guidelines in the [Contribution guide](CONTRIBUTION.md) have been run/followed
* [ ] All relevant assets have been compiled as directed in the [Contribution guide](CONTRIBUTION.md), if applicable
* [ ] All active GitHub CI checks have passed

joshcanhelp commented:
@Floppy - thanks for this! We're just about ready to release 3.11.0 (just need to merge in the remaining PR and test). Happy to try and get this in as well. 

Nothing looks too scary at this point. There are a lot of assumptions in the existing `validate_new_password` method but if those all line up with BuddyPress then we should be OK here. Error checking for the call to Auth0 is in `$this->api_change_password->call()` and that should just return `false` (IIRC) if the call fails. 

Let me know if I can help! We'll be releasing in a week or less so if we can get this in, great, if not, there's always another one around the corner :)

Floppy commented:
Thanks! This code definitely works, but I think I should look at the email change as well, and get some tests in. I've got a few days off next week so I probably won't get it done in time for 3.11, but as you say 3.12 won't be far away :)

joshcanhelp commented:
@Floppy - Just ping me when you're ready!

Floppy commented:
I've added email change support, and added some tests for the new functionality based on the old tests, but I can't quite tell if they're testing anything useful, and I'm having trouble running the test suite myself, so @joshcanhelp can you please take a look, see if I'm doing it right?

joshcanhelp commented:
@Floppy - Did you follow the steps here to setup the tests?

https://github.com/auth0/wp-auth0/blob/master/CONTRIBUTING.md#testing

If that's not working, let me know where you ran into problems so I can take a look. At first glance, it looks OK. I can pull down and try on my end ... do you see the CI checks passing? I can't see them and I think that's because this is a draft currently.

Floppy commented:
Yeah, followed those, but having trouble getting `composer test` to connect to the database. I'll start again from scratch today and see if I have more luck.

joshcanhelp commented:
Pulled down your branch and ran it locally:

```
There were 2 errors:

1) TestProfileChangeEmail::testBuddypressInitHooks
Error: Call to undefined method TestProfileChangeEmail::assertHooked()

~/Sites/wp-auth0/wp-content/plugins/auth0/tests/testProfileChangeEmail.php:63

2) TestProfileChangePassword::testBuddypressInitHooks
Error: Call to undefined method TestProfileChangePassword::assertHooked()

~/Sites/wp-auth0/wp-content/plugins/auth0/tests/testProfileChangePassword.php:94

--

There were 2 failures:

1) TestProfileChangeEmail::testSuccessfulEmailUpdateForCurrentUser
Failed asserting that false is true.

~/Sites/wp-auth0/wp-content/plugins/auth0/tests/testProfileChangeEmail.php:103

2) TestProfileChangePassword::testSuccessfulPasswordChangeDuringBuddypressProfileEdit
Failed asserting that false is true.

~/Sites/wp-auth0/wp-content/plugins/auth0/tests/testProfileChangePassword.php:186

ERRORS!
Tests: 308, Assertions: 1352, Errors: 2, Failures: 2.
```

Floppy commented:
Thanks! That's helpful :)

damieng commented:
@joshcanhelp We should push this forward or close.

joshcanhelp commented:
Agree!

@Floppy - Any news on this or should we close and revisit at another time?

Floppy commented:
Thanks for the reminder. I was working on this in the hope of getting it finished before I left Apolitical, but didn't quite make it. As far as I know though, they still really need the functionality. @Gisleburt or @JohnField may know where it stands now within that team.

joshcanhelp commented:
I'll close this for now. If this is still necessary, go ahead and re-open and we'll take a look.
